internal/pkg/cli/app_upgrade.go

```diff
@@ -0,0 +1,139 @@
+	"github.com/aws/copilot-cli/internal/pkg/deploy"
+	"github.com/aws/copilot-cli/internal/pkg/deploy/cloudformation"
+	"github.com/aws/copilot-cli/internal/pkg/describe"
+	"github.com/aws/copilot-cli/internal/pkg/term/color"
+	"github.com/aws/copilot-cli/internal/pkg/term/log"
+	termprogress "github.com/aws/copilot-cli/internal/pkg/term/progress"
+	"github.com/spf13/cobra"
+	"golang.org/x/mod/semver"
+)
+
+const (
+	fmtAppUpgradeStart    = "Upgrading application %s from version %s to version %s."
+	fmtAppUpgradeFailed   = "Failed to upgrade application %s's template to version %s.\n"
+	fmtAppUpgradeComplete = "Upgraded application %s's template to version %s.\n"
+)
+
+// appUpgradeVars holds flag values.
+type appUpgradeVars struct {
+	name string
+}
+
+// appUpgradeOpts represents the app upgrade command and holds the necessary data
+// and clients to execute the command.
+type appUpgradeOpts struct {
+	appUpgradeVars
+
+	store         store
+	prog          progress
+	versionGetter versionGetter
+	upgrader      appUpgrader
+}
+
+func newAppUpgradeOpts(vars appUpgradeVars) (*appUpgradeOpts, error) {
+	store, err := config.NewStore()
+	if err != nil {
+		return nil, fmt.Errorf("connect to config store: %v", err)
+	}
+	sess, err := sessions.NewProvider().Default()
+	if err != nil {
+		return nil, err
+	}
+	d, err := describe.NewAppDescriber(describe.NewAppDescriberConfig{})
+	if err != nil {
+		return nil, fmt.Errorf("new app describer for application %s: %v", vars.name, err)
+	}
+	return &appUpgradeOpts{
+		appUpgradeVars: vars,
+		store:          store,
+		prog:           termprogress.NewSpinner(log.DiagnosticWriter),
+		versionGetter:  d,
+		upgrader:       cloudformation.New(sess),
+	}, nil
+}
+
+// Execute updates the cloudformation stack as well as the stackset of an application to the latest version.
+// If any stack is busy updating, it spins and waits until the stack can be updated.
+func (o *appUpgradeOpts) Execute() error {
+	version, err := o.versionGetter.Version()
+	if err != nil {
+		return fmt.Errorf("get template version of application %s: %v", o.name, err)
+	}
+	if !o.shouldUpgradeApp(o.name, version) {
+		return nil
+	}
+
+	o.prog.Start(fmt.Sprintf(fmtAppUpgradeStart, color.HighlightUserInput(o.name), color.Emphasize(version), color.Emphasize(deploy.LatestAppTemplateVersion)))
+	defer func() {
+		if err != nil {
+			o.prog.Stop(log.Serrorf(fmtAppUpgradeFailed, color.HighlightUserInput(o.name), color.Emphasize(deploy.LatestAppTemplateVersion)))
+			return
+		}
+		o.prog.Stop(log.Ssuccessf(fmtAppUpgradeComplete, color.HighlightUserInput(o.name), color.Emphasize(deploy.LatestAppTemplateVersion)))
+	}()
+	app, err := o.store.GetApplication(o.name)
+	if err != nil {
+		return fmt.Errorf("get application %s: %w", o.name, err)
+	}
+	return o.upgradeApplication(app, version, deploy.LatestAppTemplateVersion)
+}
+
+func (o *appUpgradeOpts) shouldUpgradeApp(appName string, version string) bool {
```

Comment:
Why do we attach the method to the `appUpgradeOpts` struct, as opposed to just having `func shouldUpgradeApp(...)`? `o` doesn't seem to be used anywhere inside. Is it because we will need it in the future?

Reply:
Good question! This method is not meant to be reused, but yeah I can make the change.